src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,947 @@
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <time.h>
+
+#include <libyang/libyang.h>
+#include <nc_server.h>
+#include <sysrepo.h>
+
+#include "common.h"
+#include "err_netconf.h"
+#include "log.h"
+
+#define RESTORE_FILE "/restore"
+
+/**
+ * @brief Permissions for Sysrepo
+ */
+typedef enum permissions {
+    R, /* Read */
+    W  /* Write */
+} perm_t;
+
+/**
+ * @brief Context for confirmed commits.
+ *
+ * When accessing struct members must use the lock mutex.
+ * Alternatively use the set_* and get_* functions
+ */
+typedef struct commit_ctx_s {
+    char *persist;  /* What persist-id is expected */
+    timer_t timer;        /* POSIX timer used for rollback, zero if none */
+    pthread_mutex_t lock; /* Lock mutexing this structure */
+} commit_ctx_t;
+
+static commit_ctx_t commit_ctx = {.persist = NULL, .timer = NULL, .lock = PTHREAD_MUTEX_INITIALIZER};
+
+void
+ncc_commit_ctx_destroy(void)
+{
+    pthread_mutex_lock(&commit_ctx.lock);
+    if (commit_ctx.timer) {
+        timer_delete(commit_ctx.timer);
+    }
+    commit_ctx.timer = NULL;
+    free(commit_ctx.persist);
+    commit_ctx.persist = NULL;
+    pthread_mutex_unlock(&commit_ctx.lock);
+    pthread_mutex_destroy(&commit_ctx.lock);
+}
+
+static const char *
+ncc_get_persist(void)
+{
+    const char *persist;
+
+    pthread_mutex_lock(&commit_ctx.lock);
+    persist = commit_ctx.persist;
+    pthread_mutex_unlock(&commit_ctx.lock);
+
+    return persist;
+}
+
+static void
+ncc_set_persist(const char *value)
+{
+    char *new = NULL;
+
+    pthread_mutex_lock(&commit_ctx.lock);
+    free(commit_ctx.persist);
+    if (!value) {
+        commit_ctx.persist = NULL;
+        goto cleanup;
+    }
+
+    new = calloc(sizeof *new, strlen(value) + 1);
+    if (!new) {
+        EMEM;
+        goto cleanup;
+    }
+    strcpy(new, value);
+    commit_ctx.persist = new;
+
+cleanup:
+    pthread_mutex_unlock(&commit_ctx.lock);
+}
+
+static timer_t
+ncc_get_timer(void)
+{
+    timer_t timer;
+
+    pthread_mutex_lock(&commit_ctx.lock);
+    timer = commit_ctx.timer;
+    pthread_mutex_unlock(&commit_ctx.lock);
+
+    return timer;
+}
+
+/**
+ * @brief Safe setter for the timer commit_ctx_s member.
+ *
+ * @param[in] value New value for timer.
+ */
+static void
+set_timer(timer_t timer)
+{
+    pthread_mutex_lock(&commit_ctx.lock);
+    commit_ctx.timer = timer;
+    pthread_mutex_unlock(&commit_ctx.lock);
+}
+
+/**
+ * @brief Transform filename into module name.
+ *
+ * @param[in] filename Name of the file to extract module name out of.
+ *
+ * @return Module name.
+ */
+static char *
+get_module_name_from_filename(const char *filename)
+{
+    char *point = strstr(filename, ".json");
+    char *new = NULL;
+
+    if (!point) {
+        goto cleanup;
+    }
+
+    new = calloc(sizeof *new, point - filename);
+    if (!new) {
+        EMEM;
+        goto cleanup;
+    }
+
+    new = memcpy(new, filename, point - filename);
```

Comment:
Just looking at `calloc` call and `memcpy` call, it is obvious it is wrong, since the return value is `char *`, there is no terminating zero. Use `strndup(3)` instead.